Simplify `jemalloc` setup

madsmtm opened the pull request:
In the past, `#[used]` had to appear in the top-level crate to have a consistent effect on the linker. This has been fixed a while ago for ELF with the introduction of the `symbols.o` file in https://github.com/rust-lang/rust/pull/95604, and more recently for Mach-O in https://github.com/rust-lang/rust/pull/133832, which means that libraries can now implement the required workarounds themselves. This allows moving these `#[used]` declarations out of our `main.rs`.

Specifically, I have moved them into `tikv-jemalloc-sys` where they belong in https://github.com/tikv/jemallocator/pull/109 and done the same for `mimalloc` in https://github.com/purpleprotocol/mimalloc_rust/pull/146 (in case we want to experiment with switching to that one day).

Test with:
```sh
./x build library src/tools/rustdoc src/tools/clippy --set rust.jemalloc=true

# macOS
lldb -- ./build/host/stage1/bin/rustc -vV
(lldb) b _rjem_je_zone_register
(lldb) run
# Should breakpoint, this means that the allocator was properly linked

# Linux
lldb -- ./build/host/stage1/bin/rustc -vV
(lldb) b malloc
(lldb) run
# Should breakpoint, inspect that the `malloc` symbol comes from the `rustc` binary and not from `libc`
```

try-job: `aarch64-gnu`
try-job: `dist-aarch64-linux`
try-job: `dist-x86_64-musl`
try-job: `dist-x86_64-apple`
try-job: `dist-aarch64-apple`

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17762406360/job/50478094836) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50478094836) [(plain)](https://github.com/rust-lang/rust/commit/15a8fbc90c759ebc9544924c66e5762e9b30ffa0/checks/50478094836/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Caused by:
  Unable to update /jemallocator/jemalloc-sys

Caused by:
  failed to read `/jemallocator/jemalloc-sys/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:28
```

</details>

madsmtm committed:
Fix building rustdoc and clippy with jemalloc feature

madsmtm commented on compiler/rustc/src/main.rs:
```diff
@@ -36,43 +31,9 @@
 // https://github.com/rust-lang/rust/commit/b90cfc887c31c3e7a9e6d462e2464db1fe506175#diff-43914724af6e464c1da2171e4a9b6c7e607d5bc1203fa95c0ab85be4122605ef
 // for an example of how to do so.
+#[cfg(feature = "jemalloc")]
+use tikv_jemalloc_sys as _;
```
I added a paragraph to make the reason for the `use` clearer:
> NOTE: Even though Cargo passes `--extern` with `tikv_jemalloc_sys`, we still need to `use` the crate for the compiler to see the `#[used]`, see https://github.com/rust-lang/rust/issues/64402. This is similarly required if we used a crate with `#[global_allocator]`.

bors commented:
:umbrella: The latest upstream changes (presumably #148412) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

madsmtm committed:
Simplify jemalloc setup

Using the new `override_allocator_on_supported_platforms` feature in
`tikv-jemalloc-sys v0.6.1` we can avoid the manual statics.

madsmtm commented:
Thanks for the review!

@bors rollup r=jdonszelmann

Zalathar commented:
Bors, this has already been merged.

@bors r- retry

madsmtm commented:
> Some allocation changes have definitely happened there:
> 
> [snip]

Could you tell me how you arrived at that table? I'm unfamiliar with Linux perf tooling.

> I would suggest perf. testing any PR that touches things like memory allocation, for the next time 😅

Yeah, sorry about that, I should've done that!

Kobzol commented:
> Could you tell me how you arrived at that table? I'm unfamiliar with Linux perf tooling.

Yeah, sorry. If you open a benchmark row from the [comparison table](https://perf.rust-lang.org/compare.html?start=11339a0ef5ed586bb7ea4f85a9b7287880caac3a&end=d537e84520f91c06c14a8bfaad2edd601ff65287&stat=instructions:u), it shows "Command for profiling locally", which you can copy paste in a checkout of https://github.com/rust-lang/rustc-perf to do a Cachegrind diff before/after the PR.
